clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -378,6 +415,15 @@ private MemoryRecordsBuilder recordsBuilder(ByteBuffer buffer, byte maxUsableMag
     }
 
     /**
+     * Check if all batches in the queue are full.
+     */
+    private boolean allBatchesFull(Deque<ProducerBatch> deque) {
+        // Only the last batch may be incomplete, so we just check that.
+        ProducerBatch last = deque.peekLast();
+        return last == null || last.isFull();
```

Comment:
That's the only functional change in this commit, the rest is renames, comment updates and unit test fixes.

Reply:
Would it be better to add the `deque.size() > 1` condition back? Intuitively, this indicates back pressure and we should be allowed to switch away from this partition.